device/pcie/pci_device.cpp

```diff
@@ -0,0 +1,349 @@
+        write_regs(reinterpret_cast<std::uint32_t*>(static_cast<uint8_t*>(bar2_uc) + iatu_base + 0x04), &region_ctrl_2, 1);
+    }
+
+    if (device_fd != -1) {
+        ::close(device_fd);
+    }
+
+    if (bar0_wc != nullptr && bar0_wc != MAP_FAILED && bar0_wc != bar0_uc) {
+        munmap(bar0_wc, bar0_wc_size);
+    }
+
+    if (bar0_uc != nullptr && bar0_uc != MAP_FAILED) {
+        munmap(bar0_uc, bar0_uc_size);
+    }
+
+    if (bar2_uc != nullptr && bar2_uc != MAP_FAILED) {
+        munmap(bar2_uc, bar2_uc_size);
+    }
+
+    if (bar4_wc != nullptr && bar4_wc != MAP_FAILED) {
+        munmap(bar4_wc, bar4_wc_size);
+    }
+
+    if (system_reg_mapping != nullptr && system_reg_mapping != MAP_FAILED) {
+        munmap(system_reg_mapping, system_reg_mapping_size);
+    }
+
+    device_fd = -1;
+    bar0_uc = nullptr;
+    bar0_wc = nullptr;
+    bar2_uc = nullptr;
+    bar4_wc = nullptr;
+    system_reg_mapping = nullptr;
+}
+
+// Open a unique device_id per host memory channel (workaround for ttkmd < 1.21 support for more than 1 pin per fd)
+void PCIDevice::open_hugepage_per_host_mem_ch(uint32_t num_host_mem_channels) {
+    for (int ch = 0; ch < num_host_mem_channels; ch++) {
+        log_debug(LogSiliconDriver, "Opening device_fd_per_host_ch device index: {} ch: {} (num_host_mem_channels: {})", device_id, ch, num_host_mem_channels);
+        int device_fd_for_host_mem = find_device(device_id);
+        if (device_fd_for_host_mem == -1) {
+            throw std::runtime_error(std::string("Failed opening a host memory device handle for device ") + std::to_string(device_id));
+        }
+        device_fd_per_host_ch.push_back(device_fd_for_host_mem);
+    }
+}
+
+tt::ARCH PCIDevice::get_arch() const {
+    return arch;
+}
+
+template<typename T>
+T* PCIDevice::get_register_address(std::uint32_t register_offset) {
+    // Right now, address can either be exposed PCIe BAR or a TLB window
+    // Should clarify this interface
+    void *reg_mapping;
+    if (system_reg_mapping != nullptr && register_offset >= system_reg_start_offset) {
+        register_offset -= system_reg_offset_adjust;
+        reg_mapping = system_reg_mapping;
+    } else if (bar0_wc != bar0_uc && register_offset < bar0_wc_size) {
+        reg_mapping = bar0_wc;
+    } else {
+        register_offset -= bar0_uc_offset;
+        reg_mapping = bar0_uc;
+    }
+    return reinterpret_cast<T*>(static_cast<uint8_t*>(reg_mapping) + register_offset);
+}
+
+void PCIDevice::write_block(uint64_t byte_addr, uint64_t num_bytes, const uint8_t* buffer_addr) {
+    void *dest = nullptr;
+    if (bar4_wc != nullptr && byte_addr >= BAR0_BH_SIZE) {
+        byte_addr -= BAR0_BH_SIZE;
+        dest = reinterpret_cast<uint8_t *>(bar4_wc) + byte_addr;
+    }else {
+        dest = get_register_address<uint8_t>(byte_addr);
+    }
+
+    const void *src = reinterpret_cast<const void *>(buffer_addr);
+    memcpy(dest, src, num_bytes);
+// #ifndef DISABLE_ISSUE_3487_FIX
```

Comment:
For WH, I think we need to handle this.
https://yyz-gitlab.local.tenstorrent.com/tenstorrent/syseng/-/issues/3487

Reply:
I'm unable to see the ticket right now. 

Regardless, I'm a bit confused about this compile option.. I assume it's arch specific to WH? 

Also looking at metal's cmake, the `DISABLE_ISSUE_3487_FIX` option hasn't been present since mid August it seems (`702bdf9`)

Reply:
lol I'm blind, it's still in metal and only applied for GS.. the double negative is confusing so I will uncomment and change the #ifdef condition

Reply:
The fix is for WH hardware.  It's likely that Metal is not doing anything to the chip that could trigger the issue.  Can you make sure that it's applied for the WH build?  We can revisit the conditional compilation removal later.

Reply:
Yep, I have verified it's only applied for the WH build.